### PR TITLE
Add support for using new ffmpeg encoding API when writing a video 

### PR DESCRIPTION
**Pull request by peters**
In order to support encoding AV1 videos we need to use the new ffmpeg encoding API. This PR enables encoding videos using  the [rav1e](https://github.com/xiph/rav1e) encoder. 

Relates #11389 

/cc @alalek 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=linux,windows,mac,docs
```

### CODE REVIEW

File: modules/videoio/src/cap_ffmpeg_impl.hpp [hunk 1]
```diff
@@ -183,6 +183,15 @@ extern "C" {
 #endif
 #endif
 
+#ifndef USE_AV_SEND_FRAME_API
+// https://github.com/FFmpeg/FFmpeg/commit/7fc329e2dd6226dfecaa4a1d7adf353bf2773726
+#if LIBAVCODEC_BUILD >= CALC_FFMPEG_VERSION(57, 37, 100)
```
Comment (peters) [inline modules/videoio/src/cap_ffmpeg_impl.hpp:188]: @alalek Is this calculation correct?

Comment (alalek) [reply modules/videoio/src/cap_ffmpeg_impl.hpp:188]: Please add check for "FFmpeg" (to avoid libav): `LIBAVCODEC_VERSION_MICRO >= 100`

Comment (peters) [reply modules/videoio/src/cap_ffmpeg_impl.hpp:189]: Fixed by https://github.com/opencv/opencv/pull/18039/commits/6494ff6038c19e41186404d3f12b56d4a34a319a

**Comment by peters** [comment]
@alalek It appears 4 test cases are failing in the macosx build. I don't have access to a mac machine but maybe you could give me any pointers as to what might be the problem? https://pullrequest.opencv.org/buildbot/builders/precommit_macosx/builds/25298/steps/test_videoio/logs/stdio

FFmpeg documentation: https://github.com/FFmpeg/FFmpeg/blob/0842fd2c25c050e595083ba1e8c5176da3d55c9b/libavcodec/avcodec.h#L84

File: modules/videoio/src/cap_ffmpeg_impl.hpp (hunk 1 above)
Comment (peters) [reply modules/videoio/src/cap_ffmpeg_impl.hpp]: I've squashed all commits :)

**Comment by peters** [comment]
@alalek LGTM 👍

**Pull request merged**